codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/customize/RustCodegenDecorator.kt

```diff
@@ -147,26 +153,29 @@ open class CombinedCodegenDecorator<C : CoreCodegenContext>(decorators: List<Rus
                 RustCodegenDecorator::class.java,
                 context.pluginClassLoader.orElse(RustCodegenDecorator::class.java.classLoader),
             )
+                .asSequence()
+                .onEach {
+                    logger.info("Discovered Codegen Decorator: ${it.javaClass.name}")
+                }
                 // The JVM's `ServiceLoader` is woefully underpowered in that it can not load classes with generic
                 // parameters with _fixed_ parameters (like what we're trying to do here; we only want `RustCodegenDecorator`
                 // classes with code-generation context matching the input `T`).
                 // There are various workarounds: https://stackoverflow.com/questions/5451734/loading-generic-service-implementations-via-java-util-serviceloader
                 // All involve loading _all_ classes from the classpath (i.e. all `RustCodegenDecorator<*>`), and then
-                // filtering them. The most elegant way to filter is arguably by checking if we can cast the loaded
-                // class to what we want.
+                // filtering them.
+                // Note that attempting to downcast a generic class `C<T>` to `C<U>` where `U: T` is not possible to do
+                // in Kotlin (and presumably all JVM-based languages) _at runtime_. Not even when using reified type
+                // parameters of inline functions. See https://kotlinlang.org/docs/generics.html#type-erasure for details.
                 .filter {
-                    try {
-                        it as RustCodegenDecorator<T>
-                        true
-                    } catch (e: ClassCastException) {
-                        false
-                    }
+                    val clazz = T::class.java
+                    it.supportsCodegenContext(clazz)
```

Comment:
No. I thought about it and concluded I don't know how to add one. It would entail writing a test that somehow managed to modify the classpath of the JVM it's running under. Is that possible? In any case, it would then need to assert something on the received `CombinedCodegenDecorator`, but you can't reach into the object to look at the decorators that have been combined...

Reply:
Yeah, that's a fair point. I think what you can do is split the `fromClasspath` function into two halves: 1. `fromClasses` that is `internal` and has all the logic in it, and 2. `fromClasspath` which invokes the `ServiceLoader` and calls `fromClasses`. Then you can easily unit test the complex part by directly calling `fromClasses` in a unit test with a list of classes.

Reply:
Unfortunately public API inline functions are not allowed to use non-public-API declarations. See https://kotlinlang.org/docs/inline-functions.html#restrictions-for-public-api-inline-functions

But it's a small price to pay to make this unit testable. https://github.com/awslabs/smithy-rs/pull/1592/commits/a13783cff7540927c668f6590a3c3098dd1054fd